### PR TITLE
Affix search allow diacritics

### LINKED ISSUE #429: vowel length causing failed in prefix search
search "nisto-" - results in a list of forms containing "nisto-"
search "niso-" - results in a list of forms containing "nîso-"
search "nîso-" - results only in the preverb "nîso-", not forms containing it
(some further tests: "newo-" works, "nêwo-" does not)

https://sapir.artsrn.ualberta.ca/cree-dictionary/search?q=nisto-
https://sapir.artsrn.ualberta.ca/cree-dictionary/search?q=niso-
https://sapir.artsrn.ualberta.ca/cree-dictionary/search?q=n%C3%AEso-
https://sapir.artsrn.ualberta.ca/cree-dictionary/search?q=newo-
https://sapir.artsrn.ualberta.ca/cree-dictionary/search?q=n%C3%AAwo-

### PR DESCRIPTION
**Pull request by Madoshakalaka**
This fixes #429. where extra vowel length fails affix search.

The reason was the docstring of the affix search method says:

```
:arg preffix: expects lowered, no diacritics string 
```

but the caller didn't strip the diacritics before calling the function.

---
I fixed it by doing the lower-casing and diacritics-stripping inside the methods to ensure the requirement.

### CODE REVIEW

File: CreeDictionary/API/apps.py [hunk 1]
```diff
@@ -74,15 +74,9 @@ def initialize_affix_search():
     logger.info("Building tries for affix search...")
     from .models import Wordform
 
-    cree_letter_to_ascii = {
-        ascii_letter: ascii_letter for ascii_letter in string.ascii_lowercase
-    }
-    cree_letter_to_ascii.update(
-        {"â": "a", "ā": "a", "ê": "e", "ē": "e", "ī": "i", "î": "i", "ô": "o", "ō": "o"}
-    )
```
Comment (Madoshakalaka) [inline CreeDictionary/API/apps.py:82]: we already have a `remove_cree_diacritics` function and this here does the same thing.

So I extracted a function here

File: CreeDictionary/res/test_db_words.txt [hunk 2]
```diff
@@ -111,4 +111,13 @@ kîsikâw
 
 # Some paradigms of homographic entries do not show up
 nôhtêpayiw
-ayâw
\ No newline at end of file
+ayâw
+
+# some entries do not show up
+# https://github.com/UAlbertaALTLab/cree-intelligent-dictionary/issues/213
+maskwak
```
Comment (Madoshakalaka) [inline CreeDictionary/res/test_db_words.txt:118]: this is accidentally included...but should affect anything

File: CreeDictionary/utils/cree_lev_dist.py [hunk 3]
```diff
@@ -1,18 +1,35 @@
...
+    'AAEE'
+    >>> remove_cree_diacritics('âāêē')
+    'aaee'
+    >>> remove_cree_diacritics('âāêē âāêē')
+    'aaee aaee'
     """
-    nfkd_form = unicodedata.normalize("NFKD", input_str)
-    only_ascii = nfkd_form.encode("ASCII", "ignore").decode("utf-8")
-    return only_ascii
+    return "".join([_diacritic_letter_to_ascii.get(c, c) for c in input_str])
```
Comment (Madoshakalaka) [inline CreeDictionary/utils/cree_lev_dist.py:32]: replaced old magic with memorization based sweety

File: cypress/integration/regressions.spec.js [hunk 4]
```diff
@@ -242,27 +242,25 @@ context('Regressions', () => {
...
         // repeat the same test with nôhtêpayiw
         cy.visitSearch('nôhtêpayiw')
-        cy.get('[data-cy=lemma-link]').its('length').then(length=>{
+        cy.get('[data-cy=lemma-link]').its('length').then(length => {
 
-          for (let i = 0; i< length; i++){
+          for (let i = 0; i < length; i++) {
             cy.visitSearch('nôhtêpayiw')
             cy.get('[data-cy=lemma-link]').eq(i).click()
             cy.get('[data-cy=paradigm]').should('be.visible')
```
Comment (Madoshakalaka) [inline cypress/integration/regressions.spec.js:266]: eslint

File: cypress/integration/regressions.spec.js [hunk 5]
```diff
@@ -272,21 +270,28 @@ context('Regressions', () => {
         })
 
 
-
-
-
       }
-
-
```
Comment (Madoshakalaka) [inline cypress/integration/regressions.spec.js:280]: newline EXPUNGE

File: cypress/integration/regressions.spec.js [hunk 6]
```diff
@@ -272,21 +270,28 @@ context('Regressions', () => {
...
 
+    // test assumption: without vowel length, affix search works
+    cy.visitSearch('niso-')
+    cy.get('[data-cy=search-results]').should('contain', 'nîso-kîsikâw')
 
-
+    // Now if we add the vowel length, we should still get the result
+    cy.visitSearch('nîso-')
+    cy.get('[data-cy=search-results]')
+      .should('contain', 'nîso-kîsikâw')
```
Comment (Madoshakalaka) [inline cypress/integration/regressions.spec.js:294]: new integration

File: CreeDictionary/API/affix_search.py [hunk 7]
```diff
@@ -19,16 +21,19 @@ def __init__(self, words: Iterable[Tuple[str, int]]):
...
         """
-        :param suffix: expects lowered, no diacritics words
         :return: an iterable of ids
         """
+        # lower & remove diacritics
+        suffix = remove_cree_diacritics(suffix.lower())
+
         return chain(
             *[self.text_to_ids[x[::-1]] for x in self.inverse_trie.keys(suffix[::-1])]
         )
```
Comment (Madoshakalaka) [inline CreeDictionary/API/affix_search.py:39]: culprit

Comment (Madoshakalaka) [reply CreeDictionary/API/affix_search.py:39]: Don't ever trust the caller, do it yourself

File: CreeDictionary/API/affix_search.py [hunk 8]
```diff
@@ -19,16 +21,19 @@ def __init__(self, words: Iterable[Tuple[str, int]]):
 
     def search_by_prefix(self, prefix: str) -> Iterable[int]:
         """
-        :param prefix: expects lowered, no diacritics words
         :return: an iterable of ids
         """
+        # lower & remove diacritics
+        prefix = remove_cree_diacritics(prefix.lower())
```
Comment (eddieantonio) [inline CreeDictionary/API/affix_search.py:27]: Yeah, this is the right way to do it! 😂

File: CreeDictionary/res/test_db_words.txt (hunk 2 above)
Comment (eddieantonio) [reply CreeDictionary/res/test_db_words.txt:118]: That PR is approved anyway ¯\_(ツ)_/¯

File: CreeDictionary/utils/cree_lev_dist.py (hunk 3 above)
Comment (eddieantonio) [reply CreeDictionary/utils/cree_lev_dist.py:32]: You may want to look at [`str.maketrans()`](https://docs.python.org/3.3/library/stdtypes.html?highlight=maketrans#str.maketrans)/[`str.translate()`](https://docs.python.org/3.3/library/stdtypes.html?highlight=maketrans#str.translate).

File: cypress.json [hunk 9]
```diff
@@ -3,5 +3,5 @@
   "viewportHeight": 568,
   "viewportWidth": 320,
   "projectId": "8r2xra",
-  "defaultCommandTimeout": 8000
+  "defaultCommandTimeout": 15000
```
Comment (Madoshakalaka) [inline cypress.json:6]: a test case that includes typing "atchakosuk" is flaky in my local testing. It times out, which is reasonable cuz typing a bunch of stuff in succesion strains the server.

Comment (eddieantonio) [reply cypress.json:6]: yikes :/

File: cypress/integration/regressions.spec.js (hunk 6 above)
Comment (eddieantonio) [reply cypress/integration/regressions.spec.js:294]: Excellent!

**Pull request merged**